File: app/src/main/java/ch/heigvd/pro/b04/android/Poll/PollAdapter.java
package ch.heigvd.pro.b04.android.Poll;

import android.graphics.Color;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.lifecycle.LifecycleOwner;
import androidx.recyclerview.widget.RecyclerView;

import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

import ch.heigvd.pro.b04.android.Datamodel.Question;
import ch.heigvd.pro.b04.android.R;

public class PollAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder>  {
    private static final int VIEW_TYPE_HEADER = 0;
    private static final int VIEW_TYPE_QUESTION = 1;
    private static final long HEADER_ID = -1;

    private SharedViewModel state;
    private LifecycleOwner lifecycleOwner;

    private List<Question> questions = new LinkedList<>();

    public PollAdapter(SharedViewModel state, LifecycleOwner lifecycleOwner) {
        this.state = state;
        this.lifecycleOwner = lifecycleOwner;
        setHasStableIds(true);

        state.getQuestions().observe(lifecycleOwner, newQuestions -> {
            questions = newQuestions;
            Collections.sort(questions, (o1, o2) -> Double.compare(o1.getIndexInPoll(), o2.getIndexInPoll()));

            notifyDataSetChanged();
        });
    }

    @Override
    public long getItemId(int position) {
        if (position == 0)
            return HEADER_ID;

        return questions.get(position - 1).getIdQuestion();
    }

    private static class HeaderViewHolder extends RecyclerView.ViewHolder {
        private TextView title;
        public HeaderViewHolder(@NonNull ViewGroup parent, SharedViewModel state, LifecycleOwner lifecycleOwner) {
            super(LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.poll_title, parent, false));
            title = itemView.findViewById(R.id.poll_title);

            state.getPoll().observe(lifecycleOwner, poll -> {
                title.setText(poll.getTitle());
            });
        }
    }

    private class QuestionViewHolder extends RecyclerView.ViewHolder {
        private Button questionButton;

        private QuestionViewHolder(@NonNull ViewGroup parent) {
            super(LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.poll_question, parent, false));

            questionButton = itemView.findViewById(R.id.poll_question_item);
        }

        private void bindQuestion(Question question, boolean answered) {
            questionButton.setText(question.getTitle());
            questionButton.setOnClickListener(v -> state.goToQuestion(question));

            if (answered) {
                questionButton.setBackgroundColor(Color.GREEN);
            } else {
                questionButton.setBackgroundColor(Color.WHITE);
            }
        }
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        switch (viewType) {
            case VIEW_TYPE_HEADER:
                return new HeaderViewHolder(parent, state, lifecycleOwner);
            case VIEW_TYPE_QUESTION:
                return new QuestionViewHolder(parent);
            default:
                throw new IllegalStateException("Unknown view type.");
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        if(position != 0) {
            Question q = questions.get(position-1);
            ((QuestionViewHolder) holder).bindQuestion(q, q.answered() );
        }
    }

    @Override
    public int getItemCount() {
        return questions.size() + 1;
    }

    @Override
    public int getItemViewType(int position) {
        return position == 0
                ? VIEW_TYPE_HEADER
                : VIEW_TYPE_QUESTION;
    }

}

File: app/src/main/java/ch/heigvd/pro/b04/android/Poll/PollAdapter.kt
package ch.heigvd.pro.b04.android.Poll

import android.graphics.Color
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.Button
import android.widget.TextView
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.RecyclerView
import ch.heigvd.pro.b04.android.Datamodel.Question
import ch.heigvd.pro.b04.android.R
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.map
import java.util.*

class PollAdapter(private val state: PollViewModel,
                  private val lifecycleOwner: LifecycleOwner
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private var questions: List<Question> = LinkedList()
    override fun getItemId(position: Int): Long {
        return if (position == 0) HEADER_ID else questions[position - 1].idQuestion
    }

    private class HeaderViewHolder(parent: ViewGroup,
                                   state: PollViewModel,
                                   lifecycleOwner: LifecycleOwner
    ) : RecyclerView.ViewHolder(LayoutInflater.from(parent.context)
            .inflate(R.layout.poll_title, parent, false)) {

        private val title: TextView = itemView.findViewById(R.id.poll_title)

        init {
            lifecycleOwner.lifecycleScope.launchWhenStarted {
                state.poll.collect { poll -> title.text = poll.title }
            }
        }
    }

    private inner class QuestionViewHolder(parent: ViewGroup) : RecyclerView.ViewHolder(LayoutInflater.from(parent.context)
            .inflate(R.layout.poll_question, parent, false)) {

        private val questionButton: Button = itemView.findViewById(R.id.poll_question_item)

        fun bindQuestion(question: Question, answered: Boolean) {
            questionButton.text = question.title
            questionButton.setOnClickListener { state.goToQuestion(question) }
            if (answered) {
                questionButton.setBackgroundColor(Color.GREEN)
            } else {
                questionButton.setBackgroundColor(Color.WHITE)
            }
        }

    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return when (viewType) {
            VIEW_TYPE_HEADER -> HeaderViewHolder(parent, state, lifecycleOwner)
            VIEW_TYPE_QUESTION -> QuestionViewHolder(parent)
            else -> throw IllegalStateException("Unknown view type.")
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (position != 0) {
            val q = questions[position - 1]
            (holder as QuestionViewHolder).bindQuestion(q, q.answered())
        }
    }

    override fun getItemCount(): Int {
        return questions.size + 1
    }

    override fun getItemViewType(position: Int): Int {
        return if (position == 0) VIEW_TYPE_HEADER else VIEW_TYPE_QUESTION
    }

    companion object {
        private const val VIEW_TYPE_HEADER = 0
        private const val VIEW_TYPE_QUESTION = 1
        private const val HEADER_ID: Long = -1
    }

    init {
        setHasStableIds(true)
        lifecycleOwner.lifecycleScope.launchWhenStarted {
            state.questions
                    .map { it.sortedBy { question -> question.indexInPoll } }
                    .collect {
                        questions = it
                        notifyDataSetChanged()
                    }
        }
    }
}